### ISSUE TITLE
Quadratic Memory Usage (mk2)

### ISSUE THREAD

**Issue by unrealwill** [wsmoses/Enzyme#126]
The following code run fine for small n but crashes with segfault from extra memory allocations.

The code works if the integer-typed cellId is stackAllocated.
It also works if I pass an allocated cell_id as an enzyme_const parameter

`test2.cpp`
```
#include <stdio.h>
#include <iostream>
#include <stdlib.h>
#include <random>
#include <math.h>

using namespace std;

double foo( double* __restrict__ parts,int n)
 {
    const int d = 3;
    double out = 0.0;

    int* cellId = new int[n];
    double minx = 0.0;
    double maxx = 10.0;
    double h = 3.1;

    int nh = ceil((maxx-minx) / h);
    int strides[d];
    strides[0] = 1;

    for( int i = 1 ; i < d; i++)
    {
        strides[i] = strides[i-1]*nh;
    }

    for( int i = 0 ; i < n ; i++)
    {
        cellId[i] = 0;
        for( int k = 0 ; k < d ; k++)
        {
            cellId[i] += (int) ( (parts[i*d+k] - minx)/h  * strides[k]) ;
        }
    }

    for( int i = 0 ; i < n ; i++)
    {
        for( int j = 0; j < n ; j++)
        {
            double dij = 0.0;
            for( int k = 0 ; k < d ; k++)
            {
                double temp = parts[d*i+k] - parts[d*j+k];
                dij += temp*temp;
            }
            if( cellId[i] == cellId[j])
                out += dij ;

        }
    }

    delete[] cellId;
    return out;
}


int enzyme_dup;
int enzyme_out;
int enzyme_const;

typedef double (*f_ptr)(double *,int);

extern double __enzyme_autodiff(f_ptr,
    int, double *, double *,
    int, int);


int main() {

    srand(42);

    std::mt19937 e2(42);
    std::uniform_real_distribution<> dist(0, 10);
    int n = 100000;
    int d = 3;
    double* x = new double[n*d];
    double* d_x = new double[n*d];
    for( int i = 0 ; i < n*d ; i++)
    {
        x[i] = dist(e2);
        d_x[i] = 0.0;
    }

    int * cellid = new int[n];

    printf("before autodiff\n");
    __enzyme_autodiff(foo,
        enzyme_dup, x, d_x,
        enzyme_const, n);


    //printf("%f \n", y);
    for( int i = 0 ; i < 100 ; i++)
    {
    printf("dx[%d] = %f\n",i, d_x[i]);
    }

}
```
Compiled with : 
`clang test2.cpp -lstdc++ -lm -Xclang  -load -Xclang /usr/local/lib/ClangEnzyme-7.so -O2 -o test2`

**Comment by wsmoses** [comment]
So once again this code should not require any cache to be created (let alone an O(n^2) cache).

It turns out that the relevant optimization ('even if this is memory is written over, you don't need to cache it unless the potential write occurs between the place you would need to cache and its use') was already implemented in my dev branch (fortranrb) which is currently in the process of being cleaned up and merge [as of ~1.5 weeks ago it had ~40 commits of various interesting optimizations].

If you don't mind living on the bleeding edge (which by virtue of needing cleanup may be a bit more unstable) checking out the fortranrb branch should reduce addditional memory to zero. Hopefully the remainder of these optimizations should be cleaned up and merged into master within a few days.

I'm going to leave this issue open until everything is merged to master.

**Comment by unrealwill** [comment]
Thanks. Ok, I'll try to live on the bleeding edge.

**Comment by unrealwill** [comment]
I tried it on your branch and it works :+1:

**Comment by wsmoses** [comment]
Merged into master by #120
